### ISSUE TITLE
Multi tenant rule evaluation

### ISSUE THREAD

**Issue by brancz** [thanos-io/thanos#3834]
### Is your proposal related to a problem?

Scope individual rule evaluations to tenants.

### Describe the solution you'd like

Have a native solution (opt-in via a flag like `--multi-tenant`) that automatically evaluates rules scoped to a tenant. Ideally paired with #3822, as then the only thing needed for this is to set the multi tenancy header based on the tenant configured for a particular rule. A simple mean for determining the tenant could be by structuring the rule files on-disk as one directory per tenant.

### Describe alternatives you've considered

The only alternative I can think of could be to re-write rules to enforce the tenant as a label, the issue I see with this is that isolation security is an opt-in mechanism essentially, whereas users expect tenant isolation to be the default.

**Comment by bwplotka** [comment]
Discussions on Contributor Hours: 
* Wonder, how is this related to Stateless Ruler epic.

> Have a native solution (opt-in via a flag like --multi-tenant) that automatically evaluates rules scoped to a tenant. Ideally paired with #3822, as then the only thing needed for this is to set the multi tenancy header based on the tenant configured for a particular rule. A simple mean for determining the tenant could be by structuring the rule files on-disk as one directory per tenant.

How this is different to tenant label injected to rule by user? I guess it's just automation for that?

**Comment by bwplotka** [comment]
Also with https://github.com/thanos-io/thanos/issues/3822 fixed isolation would be enforced by querier on request, so the only thing might needed is just making sure the header/param is filled on ruler calling side

**Comment by brancz** [comment]
> Also with #3822 fixed isolation would be enforced by querier on request, so the only thing might needed is just making sure the header/param is filled on ruler calling side

Yes that's what I said in the issue description.

**Comment by bwplotka** [comment]
Right, makes sense to me then, thanks for proposing. LGTM, let's discuss on https://github.com/thanos-io/thanos/issues/3822 on details (header/param and external labels/tenant abstraction)

As agreed: By default non-tenant model, opt-in for build isolation and security.

**Comment by kakkoyun** [comment]
cc @Abhishek357

**Comment by yeya24** [comment]
We are still interested in this feature.

**Comment by yeya24** [comment]
Is this still valid?
